ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+
+  # Sends a cancellation request
+  def cancel() -> void
+
+  # Sends a query and awaits for the response
+  def query(name, *args) -> Any
+
+  # Sends a signal
+  def signal(name, *args) -> void
+
+  # Terminates a workflow
+  def terminate(reason, details: nil) -> void
+end
+```
+
+### Alternatives considered
+
+We thought about configuring the `Temporal::Client` with a namespace, similar to the Python SDK.
+It does bring the benefit of avoiding namespace to subsequent calls to the client. However we feel
+like it is an unjustified and artificial limitation that will impact some multi-tenant applications.
+Besides this can always be brought in the shape of a `NamespacedClient` if there's a need for it.
+
+
+## Worker
+
+The worker interface allows SDK users to start processing registered workflow and activities. These
+will then be polled for on the specified namespaces & task queues and executed.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7233')
+
+worker_1 = Temporal::Worker.new(config, 'my-namespace', 'my-task-queue')
+worker_1.register_workflow(MyWorkflow) # uses 'MyWorkflow' as a name
+worker_1.register_activity(MyActivity) # uses 'MyActivity' as a name
+worker_1.start
+
+worker_2 = Temporal::Worker.new(config, 'my-namespace', 'another-task-queue')
+worker_2.register_workflow(MyWorkflow, name: 'my-workflow') # uses 'my-workflow' as a name
+worker_2.register_activity(MyActivity, name: 'my-activity') # uses 'my-activity' as a name
+worker_2.start
+
+# later
+worker_1.stop
+worker_2.stop
+```
+
+Notes:
+
+- This is not complete example of all configuration options for a worker
+- A worker will use the workflow/activity class name as a name unless it is explicitly provided (
+  `worker_2` demonstrated this approach)
+- Each worker will run in its own thread
+- A blocking `Worker#run` method will be available for convenience of running a single worker in
+  a process until SIGTERM or SIGINT is received
+- In case connection sharing is a concern we can create and memoize a connection on the instance of
+  Temporal::Configuration
+
+And here is the `Worker` interface:
+
+```ruby
+class Worker
+  # Create a new instance of a Worker
+  def self.new(config: Temporal::Configuration, namespace: String, task_queue: String, **options) -> Worker
+
+  # Register a workflow with an optional :name
+  def register_workflow(Class, name: nil) -> void
+
+  # Register an activity with an optional :name
+  def register_activity(Class, name: nil) -> void
+
+  # Start the worker and block until the process exits
+  def run() -> void
+
+  # Start the worker asynchronously
+  def start() -> void
+
+  # Gracefully stop the worker (wait for all running workflow/activities to finish)
+  def stop() -> void
+end
+```
+
+Internally each worker will create a thread for task polling as well as a thread pool of a given
+size (passed to the initializer as an optional `:threads` argument). Then each activity task will be
```

Comment:
> a thread pool of a given size (passed to the initializer as an optional `:threads` argument)

I think you should accept the entire thread pool (can still default to creating one when not given though). It is very reasonable for me to run dozens of workers, and they should be able to share a thread pool.

Reply:
Oh, interesting… that makes perfect sense! I wonder if we should force it or have it as an option (use local thread pool unless one provided) or flip it and pick a shared thread pool from the Configuration (with an option to generate a local one). Which one do you think might be the best default approach?

Reply:
Personally I hate globals. Hate hate hate globals :-) And I consider it confusing for something like a seemingly simple configuration to have state. This is especially true when there is a shutdown path like thread pools need.

In Python we require a thread pool be provided because you can make "async activities" that don't need it, but here in Ruby every activity will need to run in a thread.

Therefore I'd say if they don't pass in a thread pool, you make a default with a reasonable max _local to that worker_.  Also, I am not familiar with Ruby, but make sure it is a max that just keeps you from having more at one time and not a count of eagerly created threads. You might be able to just make it the max out at `max_concurrent_activities` (though need to confirm with @Sushisource if we can combine that with `max_concurrent_local_activities` in some way) since you know every running activity = 1 thread.

Reply:
Agree w/ Chad here.

Core will need to be upgraded to allow some shared global max limits between workers for workflows & activities since that's been asked for once or twice. It doesn't exist now, so a pool per worker is the only thing that will work right now, but we should build with an eye towards it being shared.